File: python/448-Find All Numbers Disappeared in an Array.py
from typing import List

'''
For Space: O(1) solution
Use the trick like Leetcode442
Key: 1 ≤ a[i] ≤ n, so we can use array index to mimic our hash table
'''

class Solution:
    # def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
    #     "Hashset, Time: O(n), Space: O(n)"
    #     s = set(nums)
    #     result = []
    #     for num in range(1, len(nums)+1):
    #         if num not in s:
    #             result.append(num)
    #     return result
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        "Index trick, Time: O(2n), Space: O(1)"
        for num in nums:
            idx = abs(num) - 1
            nums[idx] = -nums[idx] if nums[idx] > 0 else nums[idx]
        result = []
        for idx, num in enumerate(nums):
            if num > 0:
                result.append(idx+1)
        return result

nums = [4,3,2,7,8,2,3,1]
result = Solution().findDisappearedNumbers(nums)
assert result == [5, 6]

File: cpp/448. Find All Numbers Disappeared in an Array.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// "Index trick, Time: O(2n), Space: O(1)"
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        for(int num : nums) {
            num = abs(num);
            int idx = num - 1;
            nums[idx] = nums[idx]<0 ? nums[idx] : -nums[idx];
        }      
        vi ans;
        for(int i=0; i<nums.size(); i++) {
            if(nums[i]>=0) ans.push_back(i+1);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; vi ans;

    solution = Solution();
    nums = {4,3,2,7,8,2,3,1};
    ans = solution.findDisappearedNumbers(nums);
    assert(ans==vi({5,6}));
}